### ISSUE TITLE
Support `devEngines` in package.json

### ISSUE THREAD

**Issue by Jiralite** [pnpm/action-setup#150]
I noticed this action has no support for `devEngines`. See [the documentation](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#devengines) for `devEngines`.

`packageManager` support should not be dropped, but it should be deprioritised. See https://github.com/nodejs/corepack/issues/567 for additional context.

**Comment by Mister-Hope** [comment]
Meeting this issue now.

We are trying to migrate to the following:

```
  "devEngines": {
    "packageManager": {
      "name": "pnpm",
      "version": "11.0.8",
      "onFail": "error"
    },
    "runtime": {
      "name": "node",
      "version": "24.15.0",
      "onFail": "error"
    }
  }
```

While the action does not care `devEngines.packageManager.version` at all.

**Comment by Mister-Hope** [comment]
@zkochan We'd better support it.

https://pnpm.io/package_json#devenginespackagemanager

I found the docs already mentioning the field during translation.

It's pretty werid that the official action does not support this while pnpm v11 already changed it's default behavior and perfers `devEngines.packageManager`. So it now seems like a bug rather than a feature request.

**Comment by Mister-Hope** [comment]
Related: https://github.com/pnpm/action-setup/issues/225 https://github.com/pnpm/action-setup/issues/227

**Comment by MindTooth** [comment]
I fixed it by changing `onFail: download`:

```json
  "devEngines": {
    "runtime": {
      "name": "node",
      "version": "26.1.0",
      "onFail": "warn"
    },
    "packageManager": {
      "name": "pnpm",
      "version": "11.0.8",
      "onFail": "download"
    }
  }
```

https://pnpm.io/package_json#devenginespackagemanager

**Comment by nemchik** [comment]
Does pnpm 10 know how to handle that? If not then everyone will have to manually update to pnpm 11, and then hopefully everything will stay consistent going forward.

**Comment by Mister-Hope** [comment]
The problem is that pnpm action is expected to prepare a supported version of pnpm. User should not be forced to set download to let node prepare it.

If node has to prepare it, then it means that the action is not doing a correct job.

However, onFail:download could be a workaround

**Comment by Mister-Hope** [comment]
> Does pnpm 10 know how to handle that? If not then everyone will have to manually update to pnpm 11, and then hopefully everything will stay consistent going forward.

Supporting devEngines in the action shall have no conflicts with keeping other behaviors.

When you run on node24 and request pnpm v10 using devEngines, the behaviors shall be similar. When you use npm, you will get msg like `this project is configured to use pnpm`

**Comment by nemchik** [comment]
> > Does pnpm 10 know how to handle that? If not then everyone will have to manually update to pnpm 11, and then hopefully everything will stay consistent going forward.
> 
> Supporting devEngines in the action shall have no conflicts with keeping other behaviors.
> 
> When you run on node24 and request pnpm v10 using devEngines, the behaviors shall be similar. When you use npm, you will get msg like `this project is configured to use pnpm`

I agree with your edit (in the action). An issue I am seeing is having `packageManager` and `devEngines.packageManager` is throwing errors and exiting non-zero with pnpm 11, so we can't have both, and `devEngines.packageManager` isn't recognized by pnpm 10, so there's no path to ensure upgrades transition smoothly, it's a hard cut. That being said, this package lacking support for `devEngines.packageManager` means we pretty much need to stick with `packageManager` even when using pnpm 11, which documents `packageManager` as legacy.

**Comment by stehefan** [comment]
Not sure who needs to hear this, but the latest version of the action supports this out of the box.

**Comment by Jiralite** [comment]
Resolved with #211. Too bad Claude did not catch it! (I hope maintainers are reading issues...)

**Comment by nemchik** [comment]
Just opened https://github.com/pnpm/action-setup/pull/273 so that the docs indicate support. Should help future users and prevent duplicate issues being opened.
